edn/src/matcher.rs

```diff
@@ -0,0 +1,114 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use std::collections::HashMap;
+use regex::Regex;
+use itertools::diff_with;
+
+use types::Value;
+
+///
+pub fn match_values<T>(pattern: T, input: T) -> bool where T: AsRef<Value> {
+    use Value::*;
+    use symbols::PlainSymbol as S;
+
+    let mut named_values: HashMap<&String, &Value> = HashMap::new();
+    let name_re = Regex::new(r"^\?[^?]+$").unwrap();
+
+    match (pattern.as_ref(), input.as_ref()) {
+        (&Nil, &Nil) => true,
+        (&Boolean(a), &Boolean(b)) => a == b,
+        (&Integer(a), &Integer(b)) => a == b,
+        (&BigInteger(ref a), &BigInteger(ref b)) => a == b,
+        (&Float(ref a), &Float(ref b)) => a == b,
+        (&Text(ref a), &Text(ref b)) => a == b,
+        (&PlainSymbol(S(ref a)), b) => {
+            if a == "_" {
+                true
+            } else if name_re.is_match(a) {
+                match named_values.insert(a, b) {
+                    None => true,
+                    Some(a) => a == b
+                }
+            } else if let &PlainSymbol(S(ref b)) = b {
+                a == b
+            } else {
+                false
+            }
+        }
+        (&NamespacedSymbol(ref a), &NamespacedSymbol(ref b)) => a == b,
+        (&Keyword(ref a), &Keyword(ref b)) => a == b,
+        (&NamespacedKeyword(ref a), &NamespacedKeyword(ref b)) => a == b,
+        (&Vector(ref a), &Vector(ref b)) => diff_with(a, b, |a, b| match_values(a, b)).is_none(),
+        (&List(ref a), &List(ref b)) => diff_with(a, b, |a, b| match_values(a, b)).is_none(),
+        (&Set(ref a), &Set(ref b)) => a.len() == b.len() && a.iter().all(|v|
+            // TOOD
+            // Should use `match_values` for this, to allow for matching rules.
+            b.contains(v)
+        ),
+        (&Map(ref a), &Map(ref b)) => a.len() == b.len() && a.iter().all(|(input_key, input_value)| {
+            // TODO
+            // This is a bit tricky because maps are unordered collections.
+            unimplemented!()
+        }),
+        _ => unimplemented!()
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    use parse;
+
+    macro_rules! assert_match_eq {
+        ( $a:tt == $b:tt, $expected:expr ) => {
```

Comment:
Interpolating the `==` like this doesn't agree with `assert_eq!`, which should be your guide.  It's misleading since you're _not_ testing equality (but matching, which is written `~=` in some languages, like Perl), and since you _can't_ do `!=` for the inversion.  Why not just `assert_matches!`, `assert_not_matches!`?